### PR TITLE
[Cosmos3] Mixed W8A8/W8A16 denoising for ModelOpt FP8 checkpoints

### PR DESCRIPTION
**Pull request by ConstBob**
# What does this PR do?

Adds denoising-step-aware mixed **W8A8/W8A16** for Cosmos3 ModelOpt FP8 checkpoints, so official Hub `revision=fp8` is not W8A8 on every step.

This is the Diffusers counterpart of the **FP8 schedule contract** in [vllm-omni#6560](https://github.com/vllm-project/vllm-omni/pull/6560) (not a port of NVFP4, tensor parallel, caches, or `vllm serve`).

### Behavior

- Checkpoint weights stay serialized static W8A8 (`weight_only: false`). Mixing is runtime.
- Video Nano / Super / Super-I2V `fp8`: first 3 + last 3 scheduler steps **W8A16** (dequant + `F.linear`); middle steps native ModelOpt **W8A8**.
- Precision is chosen once per scheduler step (CFG cond/uncond match). Reasoner defaults to A16 when the checkpoint says `reasoner: a16`.
- Distilled 4-step and Super-T2I `fp8` have `runtime: null` → mixed **off**, all steps native W8A8.
- `mixed_precision_format=None` (default) reads `quantization_config.runtime.diffusion_step_policy` from the transformer (including on-disk `transformer/config.json` when the live ModelOpt config omits `runtime`). `"none"` disables only the schedule. `"fp8"` forces it.

### Code

- `src/diffusers/pipelines/cosmos/mixed_precision.py` — policy parse (fail-closed), linear discovery, W8A16 dispatch.
- `pipeline_cosmos3_omni.py` and modular `denoise.py` — apply/reset once per step.
- `docs/source/en/api/pipelines/cosmos3.md` — user-facing FP8 mixed section.
- Out of scope: NVFP4, custom kernels, dense-weight cache, cookbook/Hub cards (follow-up after this lands).

### Tests

Unit (CPU, in-tree): `tests/pipelines/cosmos/test_cosmos3_mixed_precision.py` — **17 passed** (schedule boundaries, 1-step stays native, reasoner vs generation, overlay of live `runtime: null` + on-disk policy, FP32 W8A16, malformed policy / NVFP4 reject).

```text
pytest tests/pipelines/cosmos/test_cosmos3_mixed_precision.py
# 17 passed, 5 subtests passed
```

GPU, **not** in CI (official Hub `revision=fp8`):

- Cosmos3-Nano `@fp8`: auto-enable 3+3; 8-step T2I E2E with exact trace W8A16×3 / W8A8×2 / W8A16×3; 35-step latent W8A16×3 / W8A8×29 / W8A16×3.
- Super-Text2Image-4Step `@fp8` transformer: auto-off; GEMM stays native W8A8.

Not claimed in this PR: Super 64B generate, 189-frame I2V, vs-BF16 hashes, or the vLLM-Omni H100 throughput table.

Depends on NVIDIA ModelOpt restore for serialized FP8 (same as existing Cosmos3 FP8 loading).

### Self-review (final)

Ran the [`self-review`](https://github.com/huggingface/diffusers/blob/main/.ai/skills/self-review/SKILL.md) skill on `huggingface/main...HEAD` against `.ai/references/review-rules.md` (plus `pipelines.md`, `modular.md`, `testing.md`, `code_style.md`, `pitfalls.md`).

**Diff files:** `mixed_precision.py`, `pipeline_cosmos3_omni.py`, `modular_pipelines/cosmos/denoise.py`, `test_cosmos3_mixed_precision.py`, `docs/source/en/api/pipelines/cosmos3.md`.

**Blocking issues:** none.

**Non-blocking — left for review (deliberate):**

1. Transformer `forward` wrappers stay installed after `__call__` (`mixed_precision.py`). Inactive wrappers call the original ModelOpt forward. *Per `pipelines.md` gotcha 7 we could uninstall on exit (PAG-style); we kept wrappers to avoid re-wrapping hundreds of linears each generate.*
2. W8A16 reads `layer.weight` then `F.linear`. *Per `testing.md`, that bypasses group-offload leaf hooks.* Same dense A16 reference as vLLM-Omni; group offload + serialized FP8 was not a target.
3. Focused tests use fake ModelOpt linears + `unittest.TestCase`. *Per `testing.md`, new pipeline tests prefer real tiny classes and pytest mixins.* Existing Cosmos3 pipeline tests are still unittest; ModelOpt restore is impractical in CI.
4. `mixed_precision_overlap` is parsed from the checkpoint / `resolve()`, not exposed on `Cosmos3OmniPipeline.__call__`.
5. Docs `revision="fp8"` example does not repeat full ModelOpt `from_pretrained` kwargs; loading still follows the ModelOpt guide.

**Dead code:** `FIRST_LAST_N_FP8_POLICY`, `from_kwargs`, and `quantization_config_from_module` are used (tests + `resolve`). No unused runtime methods.

**Verdict: READY.**

## Before submitting

- [x] Did you use an AI agent (Claude Code, Codex, Cursor, etc.) to help with this PR? If so:
  - [x] Did you read the [Coding with AI agents](https://huggingface.co/docs/diffusers/main/en/conceptual/contribution#coding-with-ai-agents) guide?
  - [x] Did you run the [`self-review`](https://github.com/huggingface/diffusers/blob/main/.ai/skills/self-review/SKILL.md) skill on the diff?
  - [x] Did you share the final self-review notes in the PR description or a comment?
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/diffusers/main/en/conceptual/contribution)?
- [x] Did you read our [philosophy doc](https://huggingface.co/docs/diffusers/main/en/conceptual/philosophy)? (important for complex PRs)
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?
- [x] Are you the author (or part of the team) of the model/pipeline (only applicable for model/pipeline related PRs)?

## Who can review?

@yiyixuxu @asomoza

### CODE REVIEW

File: docs/source/en/api/pipelines/cosmos3.md [hunk 1]
```diff
@@ -43,6 +43,33 @@ Two checkpoints are released on the Hub — [`nvidia/Cosmos3-Nano`](https://hugg
...
+result = pipe(prompt="...", num_inference_steps=35)  # 3×W8A16 / 29×W8A8 / 3×W8A16
+```
+
+Call-site overrides:
+
+- `mixed_precision_format="none"` disables the schedule only (quantized W8A8 remains).
+- `mixed_precision_format="fp8"` forces the first/last-N schedule even if the checkpoint has no policy.
+- `mixed_precision_first_steps` / `mixed_precision_last_steps` / `mixed_precision_reasoner_policy` override the checkpoint counts and reasoner path (`"high_precision"` = W8A16, `"base_precision"` = native W8A8).
+
+These kwargs are not Accelerate `mixed_precision`. They only select W8A8 vs W8A16 on Cosmos3 ModelOpt linears.
```
Comment (sayakpaul) [inline docs/source/en/api/pipelines/cosmos3.md:71]: I don't think we need to mention things like these. If we want the users to control them, we could simplify the text.

File: docs/source/en/api/pipelines/cosmos3.md [hunk 2]
```diff
@@ -43,6 +43,33 @@ Two checkpoints are released on the Hub — [`nvidia/Cosmos3-Nano`](https://hugg
 > [!TIP]
 > Make sure to check out the Schedulers [guide](../../using-diffusers/schedulers) to learn how to explore the tradeoff between scheduler speed and quality, and see the [reuse components across pipelines](../../using-diffusers/loading#reusing-models-in-multiple-pipelines) section to learn how to efficiently load the same components into multiple pipelines.
 
+## FP8 mixed W8A8/W8A16 denoising
+
+Official ModelOpt FP8 checkpoints live on the Hub `fp8` revision (for example [`nvidia/Cosmos3-Nano`](https://huggingface.co/nvidia/Cosmos3-Nano) with `revision="fp8"`). The serialized weights are static W8A8. Video Nano / Super / Super-I2V checkpoints also store a `quantization_config.runtime.diffusion_step_policy` on the transformer: the **first 3 and last 3** denoising steps run **W8A16** (dequantized FP8 weights, `torch.nn.functional.linear`), and the middle steps keep native **W8A8**. Precision is chosen once per scheduler step so CFG cond/uncond calls match. Distilled 4-step and Super-T2I FP8 checkpoints omit that policy (`runtime` is `null`) and stay native W8A8 on every step.
```
Comment (sayakpaul) [inline docs/source/en/api/pipelines/cosmos3.md:48]: Users may not be familiar with the convention of W8A8. It would make sense to elaborate on that.

File: tests/pipelines/cosmos/test_cosmos3_mixed_precision.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
+# Copyright 2026 The NVIDIA Team and The HuggingFace Team. All rights reserved.
```
Comment (sayakpaul) [inline tests/pipelines/cosmos/test_cosmos3_mixed_precision.py:1]: We don't need these tests yet. Let's first gauge if this checkpoint is used much.

File: src/diffusers/pipelines/cosmos/pipeline_cosmos3_omni.py [hunk 4]
```diff
@@ -1685,98 +1708,77 @@ def __call__(
...
-                    action_condition_mask=[action_condition_mask] if action_condition_mask is not None else None,
-                    raw_action_dim=raw_action_dim_resolved,
-                )
+        mixed_precision = Cosmos3MixedPrecisionConfig.resolve(
+            self.transformer,
+            mixed_precision_format=mixed_precision_format,
+            mixed_precision_first_steps=mixed_precision_first_steps,
+            mixed_precision_last_steps=mixed_precision_last_steps,
+            mixed_precision_reasoner_policy=mixed_precision_reasoner_policy,
+        )
```
Comment (sayakpaul) [inline src/diffusers/pipelines/cosmos/pipeline_cosmos3_omni.py:1717]: How does it work out when people would use a different quantization backend like TorchAO, for example?

**Review by sayakpaul — COMMENTED** [review]
Thanks for the PR. I think it's a bit intrusive in terms of what it is doing. Could you provide runtime numbers (latency, throughput) and also how does it compare in terms of visual results (with and without comparisons would be great).

File: docs/source/en/api/pipelines/cosmos3.md (hunk 1 above)
Comment (ConstBob) [reply docs/source/en/api/pipelines/cosmos3.md]: Agreed. The guide now only explains the checkpoint default and `mixed_precision_format="none"`. The other kwargs stay on `__call__` for debugging but are no longer listed in the user docs.

File: docs/source/en/api/pipelines/cosmos3.md (hunk 2 above)
Comment (ConstBob) [reply docs/source/en/api/pipelines/cosmos3.md]: Done. The section now defines W8A8 vs W8A16.

File: tests/pipelines/cosmos/test_cosmos3_mixed_precision.py (hunk 3 above)
Comment (ConstBob) [reply tests/pipelines/cosmos/test_cosmos3_mixed_precision.py]: Removed in the latest commit.

File: src/diffusers/pipelines/cosmos/pipeline_cosmos3_omni.py (hunk 4 above)
Comment (ConstBob) [reply src/diffusers/pipelines/cosmos/pipeline_cosmos3_omni.py:1714]: It is a no-op for other backends. Mixed W8A8/W8A16 only applies to serialized ModelOpt FP8. TorchAO, bitsandbytes, and unquantized weights resolve to `format="none"` and we do not wrap their linears. Forcing `mixed_precision_format="fp8"` on a non-ModelOpt backend logs a warning and still leaves the native forward unchanged.

**Comment by ConstBob** [comment]
Thanks @sayakpaul . I've updated the PR per your comments (please refer to the replies to your each comments).

With vs without on `nvidia/Cosmos3-Nano` `@fp8` (same weights; UniPC `flow_shift=10`, guidance 6, seed 0, 720×1280, 35 steps; 1 warmup + 3 timed, median; RTX PRO 6000 Blackwell). Without = `mixed_precision_format="none"` (all W8A8, not BF16). Mixed follows the checkpoint: W8A16×3 / W8A8×29 / W8A16×3. Times below are full `pipe()` wall clock; parentheticals are the three timed runs in order.
| | T2I 720×1280 35-step | T2V 49f 720p 35-step (~2 s @ 24 fps) |
|---|---|---|
| mixed latency | 6.858 s (7.025 / 6.858 / 6.844) | 68.812 s (68.369 / 68.812 / 68.965) |
| none latency | 5.399 s (5.345 / 5.399 / 5.405) | 60.881 s (60.881 / 61.191 / 59.525) |
| generation throughput (mixed vs none) | 0.146 vs 0.185 img/s | 0.712 vs 0.805 frame/s |
| mixed vs none | +27% | +13% |

Images/videos attached (mixed vs `none`).

`t2i_mixed.jpg`:

t2i_mixed

`t2i_none.jpg`:

t2i_none

`t2v_49f_mixed.mp4`:

https://github.com/user-attachments/assets/7e296cdb-56db-4623-9c4b-7155c14d2a2c

`t2v_49f_none.mp4`:

https://github.com/user-attachments/assets/8e4a1380-6deb-4a94-9b4f-de7d8b254a6f

File: docs/source/en/api/pipelines/cosmos3.md [hunk 5]
```diff
@@ -43,6 +43,36 @@ Two checkpoints are released on the Hub — [`nvidia/Cosmos3-Nano`](https://hugg
...
+Image generation and few-step distilled checkpoints do not show that flickering, so Super-T2I and the distilled 4-step FP8 repos declare no schedule and stay W8A8 on every step. The schedule is also **ModelOpt FP8 only**: other quantization backends (for example TorchAO) keep their native forwards.
+
+Loading follows the [ModelOpt guide](../../quantization/modelopt): call `enable_huggingface_checkpointing()` so the serialized quantizers are restored. The `fp8` revision already carries its quantization config, so you do not pass one, and mixed precision follows the checkpoint automatically:
+
+```python
+import torch
+from diffusers import Cosmos3OmniPipeline
+from modelopt.torch.opt import enable_huggingface_checkpointing
+
+enable_huggingface_checkpointing()
```
Comment (sayakpaul) [inline docs/source/en/api/pipelines/cosmos3.md]: Why do we need this?

File: docs/source/en/api/pipelines/cosmos3.md [hunk 6]
```diff
@@ -43,6 +43,36 @@ Two checkpoints are released on the Hub — [`nvidia/Cosmos3-Nano`](https://hugg
...
+pipe = Cosmos3OmniPipeline.from_pretrained(
+    "nvidia/Cosmos3-Nano",
+    revision="fp8",
+    dtype=torch.bfloat16,
+    device_map="cuda",
+)
+result = pipe(prompt="...", num_inference_steps=35)
+```
+
+Pass `mixed_precision_format="none"` to keep every step on native W8A8.
```
Comment (sayakpaul) [inline docs/source/en/api/pipelines/cosmos3.md]: Can we discuss the options available and the trade-offs amongst them?

**Review by sayakpaul — COMMENTED** [review]
Thanks for the iteration. Do you think it could make sense to try to think of an approach where we make the changes bit more general, i.e., we make it so that these are done at the Modelopt quantizer class?

https://github.com/huggingface/diffusers/blob/a71e62e0d226c284b86abf518791a5ffbba064bf/src/diffusers/quantizers/modelopt/modelopt_quantizer.py#L28

Or do you think the pipeline-level call arguments make it difficult to do so?

File: docs/source/en/api/pipelines/cosmos3.md (hunk 5 above)
Comment (ConstBob) [reply docs/source/en/api/pipelines/cosmos3.md]: Dropped from this snippet. Serialized ModelOpt restore stays in the ModelOpt guide; mixed precision only needs the fp8 revision once the transformer is loaded.

File: docs/source/en/api/pipelines/cosmos3.md (hunk 6 above)
Comment (ConstBob) [reply docs/source/en/api/pipelines/cosmos3.md]: Expanded the default vs "none" trade-off (video flicker vs the extra cost of dequant + F.linear).

**Comment by ConstBob** [comment]
Thanks @sayakpaul for the review. I've updated the PR. 

I looked at the other Cosmos3 FP8 runtimes that already landed this schedule. None of them put it on the generic ModelOpt loader/quantizer.

vLLM-Omni (vllm-project/vllm-omni#6560, merged) keeps policy parse + W8A16 dispatch under `diffusion/models/cosmos3/mixed_precision/`. The pipeline selects precision at each denoising-loop iteration; generate-time disable is `--additional-config '{"cosmos3_mixed_precision":{"enabled":false}}'`.

SGLang (sgl-project/sglang#36380, merged) does the same from the Cosmos3 denoise stage (`StepMixedPrecisionController` + dequant/`F.linear`). The generic FP8 linear method is unchanged; an env var disables only the schedule.

That matches this PR: `NVIDIAModelOptQuantizer` is load-time (restore FP8 weights), while the schedule is a function of scheduler step index, CFG, and reasoner vs generation. The `__call__` kwargs are the Diffusers equivalent of vLLM's additional-config / SGLang's env — request-scoped, not `from_pretrained`.

To align with this PR's counterparts in other runtimes, I'd prefer to keep that split in this PR rather than moving the schedule onto `NVIDIAModelOptQuantizer`. Does that sound reasonable to you?

**Review by sayakpaul — APPROVED** [review]

**Comment by sayakpaul** [comment]
@bot /style
